### ISSUE TITLE
Constant terms in Hamiltonian related to external parameters

### ISSUE THREAD

**Issue by ahttraga** [scqubits/scqubits#179]
For custom circuits, constant terms in the Hamiltonian associated with external parameters are subtracted before calculating energies.  It is then difficult to compare the eigenenergies for different values of these parameters.  For example, the offset charge dispersion of a CPB implemented as a custom circuit has a parabolic background.

### Documentation check
The bug results in unexpected behavior
 - [X] I have checked the API documentation.
 - [X] I could not locate relevant information in the documentation or information is missing. 

### Reproduce the bug
1. Implement a CPB as a custom circuit with a single `JJ` line and $E_J = 1, E_C = 1/4$. Don't use the transmon class.
2. Plot the eigenenergies as a function of $n_{g1}$.
3. Observe an overall negative parabolic background.
4. Checking `_hamiltonian_sym_for_numerics` shows that the $n_{g1}^2$ term is missing.

### Expected behavior
There should be no parabolic background in the spectrum.  In the transmon class, the offset charging term is [incorporated in the Hamiltonian](https://github.com/scqubits/scqubits/blob/4aa1474adc73caf9e7399e0cb8d4ad4564f78351/scqubits/core/transmon.py#L120), giving the correct spectrum (as in one of the examples).

### Code
The background corresponds to a quadratic term in $n_{g1}$ which I think is [subtracted from the Hamiltonian as an irrelevant constant](https://github.com/scqubits/scqubits/blob/4aa1474adc73caf9e7399e0cb8d4ad4564f78351/scqubits/core/circuit.py#L1006).  

### Fix
As a hack, I commented the line above and was able to reproduce the correct CPB spectrum.  The external flux offset energies may also need to be fixed for circuits with inductances.

### OS and version used (please complete the following information):
 - OS: MacOS
 - scqubits version 3.1.1
 - Python version 3.11

**Comment by jkochNU** [comment]
Thank you for filing this! My broad take on this topic has always been that c-number offsets to superconducting-circuit Hamiltonians do not carry meaning and should be ignored when comparing numerical values. Note:
* The transmon Hamiltonian "should" have a potential energy term of + EJ * (1 - cos phi), yet the +EJ contribution is most commonly omitted.
* The compact kinetic form 4EC(n-ng)^2 is only obtained by allowing for the addition of a c-number term.

In other words, c-number offsets in H (which lead to a global shift of the entire energy spectrum) are arbitrary. What counts are energy differences (i.e., transition energies). 

@ahttraga I'd like to understand better whether you are indicating that there is a deeper problem we may be missing.

**Comment by ahttraga** [comment]
Thank you for making this package.  It's excellent!  Finally we can go from netlist to spectrum!

The problem is that the energy levels as a function of charge offset are not consistent between the Transmon qubit and an equivalent custom circuit.

If I make a custom circuit with a JJ yaml such that 'Ej = 1, Ec = 1/4', here are the energy levels as a function of 'ng1' using 'plot_evals_vs_paramvals'

<img width="491" alt="image" src="https://user-images.githubusercontent.com/54644739/233375417-ca252331-9370-4c54-b260-c440465bc694.png">

This is because the corresponding '_hamiltonian_sym_for_numerics' is '-E_J*cos(θ1) + 4.0*E_C*n1**2 + 8.0*E_C*I*n1*ng1' which doesn't have an 'ng1**2' term.  This plot is not what I expect to see for a CPB.

The equivalent circuit implemented as a Transmon class has '(n1-ng1)^2' along the diagonal of the Hamiltonian, so you don't have the parabolic background in the plot and it looks correct. I can also obtain the correct result for the custom circuit if I comment out the line of code I mentioned, such that '_hamiltonian_sym_for_numerics' is '-E_J*cos(θ1) + 4.0*E_C*n1**2 + 4.0*E_C*I**2*ng1**2 + 8.0*E_C*I*n1*ng1', with the 'ng1**2' term.

So I agree with your comments in theory, but in practice I think the two plots should be the same.  For consistency in general, you may want to include the c-numbers in the custom circuit class that correspond to parameters such as flux or charge offset that are implicitly included in the Hamiltonians of the Qubit classes.  Of course there's no problem for those who only plot transition energies or wavefunctions.

**Comment by jkochNU** [comment]
@ahttraga my point is that it is debatable that one or the other is "correct". The ng^2 terms gets artificially added in the first place to complete the square@ahttraga my point is that it is debatable that one or the other is "correct". The ng^2 terms gets artificially added in the first place to complete the square in standard circuit quantization, if I recall correctly. I would gently push against any interpretation that assigns meaning to absolute energies, as those are not measurable quantities.
Comparisons should be made for ground energy subtracted (and `plot_evals_vs_paramvals` offers that option).  in standard circuit quantization, if I recall correctly. I would gently push against any interpretation that assigns meaning to absolute energies, as those are not measurable quantities.
Comparisons should be made for ground energy subtracted (and `plot_evals_vs_paramvals` offers that option).

**Comment by ahttraga** [comment]
Perhaps I'm mistaken but I find that the Hamiltonian for a CPB, obtained from the standard Lagrangian, already has the ng^2 term needed to complete the square.  And I believe that the symbolic Hamiltonian obtained by scqubit for a "custom circuit" Josephson junction also has this term: (output of `jj.sym_hamiltonian()`)

$$
\left(4.0 E_{C} n_{1}^{2} + 4.0 E_{C} n_{g1}^{2} + 8.0 E_{C} n_{1} n_{g1}\right) - E_{J} \cos{\left(θ_{1} \right)}
$$

whereas `jj_hamiltonian_sym_for_numerics` gives

`-E_J*cos(θ1) + 4.0*E_C*n1**2 + 8.0*E_C*I*n1*ng1`

showing that the `ng1**2` term has been subtracted.  Whether it is correct or not to subtract, it's up to you, but this default behavior for the custom circuit class results in an inconsistency with the Transmon qubit class.

**Comment by ahttraga** [comment]
I agree in principle.  Again, I was just surprised that the `Circuit` class [CPB charge dispersion](https://github.com/scqubits/scqubits/issues/179#issuecomment-1516301968) was different from the `Transmon` dispersion for the same circuit.  A statement in the documentation that the `Circuit` class removes constant terms from the Hamiltonian, including those related to external parameters, would have avoided this confusion.

**Comment by saipavanc** [comment]
@petergthatsme After reading your comment and also talking with Jens, I did find the inconsistency in Circuit module. It might have come about in my previous attempts to deal with constants when Hierarchical diagonalization is used. I just fixed it in the commit 04cb203, where the symbolic Hamiltonian is now consistent with the numerical Hamiltonian.

@ahttraga This also removes the parabolic background in the eigenvalues from the Circuit module, as all the constants are taken into account. But as others have already mentioned, this is just a global shift of energies which is not a physical observable.

![image](https://user-images.githubusercontent.com/79005467/234399504-1aeb3a2e-5867-4639-940d-bcf11f5f87e3.png)

**Comment by petergthatsme** [comment]
great - thanks for the quick fix @saipavanc 

I will close this issue for now, but @ahttraga  if other stuff comes up or something is still not right, please don't hesitate to reopen (also Sai's fix will make it into the next formal release, of course).
